### ISSUE TITLE
HuBERT positional convolution is sensitive to padding when conv_pos_batch_norm=True

### ISSUE THREAD

**Issue by Abdulrhman-shaheen** [huggingface/transformers#47739]
I noticed a potential padding-handling issue in HuBERT when `conv_pos_batch_norm=True`.

With this configuration, `HubertPositionalConvEmbedding` applies `BatchNorm1d` without accounting for the padded time steps. 

See [modeling_hubert: L85](https://github.com/huggingface/transformers/blob/main/src/transformers/models/hubert/modeling_hubert.py#L85)

This has two consequences:

1. Padding positions no longer remain zero after batch normalization.
2. [only in training mode] The normalized representations of valid time steps are influenced according to the padding.

As a result, the output for a given sample may change depending on the lengths of the other samples in the same batch, even when the sample itself is unchanged.

I believe this behavior might be in other models as well. So it's worth checking for a similar pattern.

A quick fix would be to at least zero out the padded time frames again after normalization by passing the attention mask, but the second consequence still remains.

Here's a quick example with output that shows the difference:

```python
import torch
from transformers import HubertConfig
from transformers.models.hubert.modeling_hubert import (
    HubertPositionalConvEmbedding,
)
from torch.nn.utils.rnn import pad_sequence

config = HubertConfig(
    hidden_size=1,
    num_conv_pos_embeddings=3,
    num_conv_pos_embedding_groups=1,
    feat_extract_activation="linear",
    conv_pos_batch_norm=True,
)

layer = HubertPositionalConvEmbedding(config).eval()

with torch.no_grad():
    layer.batch_norm.bias.normal_() # bias is initialized to 0, so we need to set it to something else to see the effect of batch norm

sz = torch.randint(1, 10, (1,)).item()

sample1 = torch.randn(sz, 1)
sample2 = torch.randn(2 * sz, 1) # Longer sequence to pad the first sample to.

batch = pad_sequence([sample1, sample2], batch_first=True)

output_alone = layer(sample1.unsqueeze(0))
output_batched = layer(batch)[0, :len(sample1)]

print(output_alone)
print(output_batched)
print((output_alone - output_batched).abs().mean().item())  # Non-negligible difference.
```

**Comment by tomerzipori** [comment]
Hi! jumping into this, played around a little with it now. I think the optimal solution would be to somehow apply the attention mask to ignore padding before batchnorm, and add it immediately after (at training time). This will both keep mean and variance estimates clean, and the padding's zeros will remain zeros after it.
If we don't want to wait for a masked nn.BatchNorm1d from PyTorch, I guess it can be implemented here as some masked_batchnorm that wraps nn.BatchNorm1d with stripping padding and adding them back after the norm.

**Comment by tomerzipori** [comment]
Ran a downstream experiment to test whether this padding sensitivity affects intent-classification performance. Used some 

### Setup

- Model: `ntu-spml/distilhubert`, pinned at `fa87d96265d6b7af66e112faff6ff44df419cec9`
- Dataset: English MInDS-14 (`en-AU`, `en-GB`, and `en-US`)
- Task: 14-class intent classification
- Test set: 274 examples
- Primary metric: test macro-F1 from the best validation checkpoint

I used a 2 × 2 design:

| Condition | Positional normalization | Training padding |
|---|---|---|
| C-L | Current `BatchNorm1d` | Minimal batch padding |
| C-H | Current `BatchNorm1d` | Fixed 20-second tensors |
| F-L | Masked BatchNorm over valid frames | Minimal batch padding |
| F-H | Masked BatchNorm over valid frames | Fixed 20-second tensors |

The valid audio, split, batch membership and order, initialization, optimizer, and number of updates were held fixed within each seed.

### Main result

When all models were evaluated with fixed 20-second tensors:

| Method | Low-padding training | High-padding training | Change |
|---|---:|---:|---:|
| Current BatchNorm | 0.4413 | 0.4220 | -0.0192 |
| Masked BatchNorm | 0.4536 | 0.4542 | +0.0006 |

The paired difference-in-differences was:

`(F-H - F-L) - (C-H - C-L) = +0.0198 [-0.0407, +0.0803] macro-F1`

The experiment does not establish that masked BatchNorm will improve every model, or even every HuBERT model at this task. It used one compact DistilHuBERT checkpoint, one English intent dataset, and five seeds.

<img width="1972" height="644" alt="Image" src="https://github.com/user-attachments/assets/fbb6ef84-b678-4395-90c7-0dda60becacf" />

I can share the raw data, raw results, and analysis script if useful.

### Masked BatchNorm implementation

The specific masked batchnorm I tested here goes conceptually like this:
For an input with shape `[batch, channels, time]`,

1. Used HuBERT's feature-level attention mask.
2. Gathered all valid frames across the batch and time dimensions into a two-dimensional `[valid_frames, channels]` tensor.
3. Applied the existing `nn.BatchNorm1d` to that tensor.
4. Scattered the normalized frames back to their original positions.
5. Set every padded position to exactly zero before the positional convolution.

In simplified form:

```python
feature_first = hidden_states.transpose(1, 2)
valid_states = feature_first[feature_mask]
normalized_valid_states = batch_norm(valid_states)

normalized = torch.zeros_like(feature_first)
normalized[feature_mask] = normalized_valid_states

**Comment by YeonwooSung** [comment]
Thanks for the report, and thanks @tomerzipori for the 2026-09-10 experiment. On current main, `HubertPositionalConvEmbedding` still applies unmasked `BatchNorm1d` when `conv_pos_batch_norm=True`, and the wav2vec2 encoder zeros pads then drops the 2D mask before `pos_conv_embed`. The default is `False`, so `facebook/hubert-base` is unaffected.

#47765 already tried a fix and was closed. Two design questions before another PR:

1. Fairseq `make_conv_pos` also uses unmasked `BatchNorm1d` after zeroing pads. Changing this would be a numerical break for `conv_pos_batch_norm=True` checkpoints (e.g. mHuBERT-25Hz) on padded batches. Is that intended, or should we keep fairseq parity?
2. If we do change it, should we gather/scatter valid frames through the existing `nn.BatchNorm1d` (as @tomerzipori sketched), HuBERT-only? wav2vec2 does not have this flag.

@Abdulrhman-shaheen @tomerzipori is either of you already working on a PR?

**Comment by tomerzipori** [comment]
@YeonwooSung Not currently working on one, would be happy to once we agree on the optimal design.

**Comment by YeonwooSung** [comment]
Thanks — I'll wait for a maintainer call on fairseq parity vs masked BN before anyone opens a PR.
